proxy/coco/src/peer.rs

```diff
@@ -261,6 +274,24 @@ impl Api {
         Ok(project_meta)
     }
 
+    /// Query the network for providers of the given [`RadUrn`].
+    ///
+    /// Notes:
+    ///   * there is no guarantee that a peer who claims to provide the [`RadUrn`] actually has it,
+    ///     nor that it is reachable using any of the addresses contained in [`PeerInfo`].
+    ///
+    ///   * there is no guarantee that any peer ever replies to the query, meaning that awaiting for
+    ///     a reply might never terminate, requiring callers of this function to timeout on
+    ///     `next()`.
+    #[warn(clippy::future_not_send)]
+    pub async fn providers(
+        &self,
+        urn: RadUrn,
+    ) -> BoxFuture<'static, impl futures::Stream<Item = PeerInfo<IpAddr>>> {
```

Comment:
```suggestion
    ) -> impl Futue<Output = futures::Stream<Item = PeerInfo<IpAddr>>> {
```

What's the reason that we can't return an `impl` here?

Reply:
Don't get it either -- the desugared return type is 

```rust
impl Future<Output = Pin<Box<dyn Future<Output = Stream<Item = PeerInfo<IpAddr>> + 'a + Send>>>
``` 
Double future!

Reply:
:rainbow: :rainbow:

Reply:
I am confused :upside_down_face: 

From this comment (https://github.com/radicle-dev/radicle-upstream/pull/875#discussion_r484346892), as well as https://github.com/radicle-dev/radicle-upstream/pull/875#discussion_r484342112, and given the warnings we were getting, I understood that the way to address the issue was to wrap the stream in a `BoxFuture`.

> What's the reason that we can't return an impl here?

Because [`Stream` is not `Send`](https://buildkite.com/monadic/radicle-upstream/builds/4515#e234d536-6616-4279-b94d-efc23e8094ea/626-634), if I don't misunderstand the issue.

Reply:
The thing is that `async fn` desugars to something which is a `Future` and captures `self` (I don't know exactly how, but it's good enough to think about it like this). Minimally, `self` has to be `Send`, so it can be moved to the executor threadpool. Most of the time, however, it also has to be `Sync`, which `PeerApi` isn't.

The trick employed in `PeerApi` is that the returned `Future` does **not** capture `self`, but a clone of `Protocol`, which _is_ `Sync`. Thus, if you just return the `Future` returned from `PeerApi::providers`, all should be good.

It would also be possible to `.await` that future and return just the `Stream`, but for this to work, `self.peer_api` would need to be mutexed by an _async aware_ mutex (e.g. `futures::lock::Mutex`).

Reply:
@kim, thanks a ton for such an insightful explanation. I got it, pushed :)